### PR TITLE
Write the first half of the advanced async/await chapter

### PR DESCRIPTION
**Pull request by nrc**

### CODE REVIEW

**Comment by traviscross** [comment]
cc @rust-lang/wg-async

**Comment by traviscross** [comment]
I notice that the recent PRs have been merged with the CI red.  The first breaking one seems to be 58eeade3ecdee7b250330ce2b75f7e0222ca552e.  What's the plan here?

File: src/part-guide/adv-async-await.md [hunk 1]
```diff
@@ -2,26 +2,142 @@
...
+async {
+    let x = foo()?;   // This `?` only exits the async block, not the surrounding function.
+    consume(x);
+    Ok(())
+}.await?
+```
+
+Annoyingly, this often confuses the compiler since (unlike functions) the 'return' type of an async block is not explicitly stated. You'll probably need to add some type annotations on variables or use turbofish types to make this work, e.g., `Ok::<_, MyError>(())` instead of `Ok(())` in the above example.
+
+A function which returns an async block is pretty similar to an async function. Writing `async fn foo() -> ... { ... }` is roughly equivalent to `fn foo() -> ... { async { ... } }`. In fact, from the caller's perspective they are equivalent, and changing from one form to the other is not a breaking change. Furthermore, you can override one with the other when implementing an async trait (see below). However, you do have to adjust the type, making the `Future` explicit in the async block version: `async fn foo() -> Foo` becomes `fn foo() -> impl Future<Output = Foo>` (you might also need to make other bounds explicit, e.g., `Send` and `'static`).
```
Comment (traviscross) [inline src/part-guide/adv-async-await.md:114]: In Rust 2024, due to the changes to lifetime capture rules, these should be exactly equivalent.  In either case, auto traits like `Send` are going to leak to the returned opaque future.

It'd probably be more accurate to say that there are times when you want or need to desugar to the `-> impl Future<..>` form so that you can specify bounds on that returned opaque type rather than saying that one needs to adjust those things when converting between the forms.

**Comment by traviscross** [comment]
This looks great overall.

File: src/part-guide/adv-async-await.md [hunk 2]
```diff
@@ -2,26 +2,142 @@
...
+
+We'll talk soon about non-blocking/async IO. For now, just know that non-blocking IO is IO which the async runtime knows about and so will only block the task which is waiting for it, not the whole thread. It is very important to only use non-blocking IO from an async task, never blocking IO (which is the only kind provided in Rust's standard library).
+
+### Blocking computation
+
+You can also block the thread by doing computation (this is not quite the same as blocking IO, since the OS is not involved, but the effect is similar). If you have long-running computation (with or without blocking IO) without yielding control to the runtime, then that task will never give the scheduler a chance to schedule other tasks. Remember that async programming uses cooperative multitasking? Here a task is not cooperating, so other tasks won't get a chance to get work done. We'll discuss ways to mitigate this later.
```
[the remark this answers is not in the document]
Comment (nrc) [reply src/part-guide/adv-async-await.md:31]: I've introduced the scheduler as part of the runtime in the previous chapter, but I haven't really talked about waking yet, so I think I prefer the scheduling terminology here.

File: src/part-guide/adv-async-await.md [hunk 3]
```diff
@@ -2,26 +2,142 @@
...
+
+There are many other ways to block a whole thread, and we'll come back to blocking several times in this guide.
+
+### Cancellation
+
+Cancellation means stopping a future (or task) from executing. Since in Rust, futures must be driven forward by an external force (like the async runtime), if a future is no longer driven forward then it will not execute any more. If a future is dropped (remember, a future is just a plain old Rust object), then it can never make any more progress and is cancelled.
```
[the remark this answers is not in the document]
Comment (nrc) [reply src/part-guide/adv-async-await.md:37]: Makes sense in general, in this case though I think the first 'in Rust' is warranted because I want this is a behaviour which is pretty unique to Rust and in contrast to other systems (and this is a known footgun for those coming from other languages).

The second use is a bit of an in-joke referencing POJOs: https://en.wikipedia.org/wiki/Plain_old_Java_object

File: src/part-guide/adv-async-await.md [hunk 4]
```diff
@@ -2,26 +2,142 @@
...
+An example of how this can go wrong is if an async function reads data into an internal buffer, then awaits the next datum. If reading the data is destructive (i.e., cannot be re-read from the original source) and the async function is cancelled, then the internal buffer will be dropped, and the data in it will be lost.
+
+We'll be coming back to cancellation and cancellation safety a few times in this guide, and there is a whole [chapter]() on the topic in the reference section.
+
+
+## Async blocks
+
+A regular block in Rust (`{ ... }`) groups code together in the source and creates a scope of encapsulation for names. At runtime, the block is executed in order and evaluates to the value of it's last expression (or void (`()`) if there is no trailing expression).
+
+Similarly to async functions, an async block is a deferred version of a regular block. An async block scopes code and names together, but at runtime it is not immediately executed and evaluates to a future. To execute the block and obtain the result, it must be `await`ed. E.g.,
```
[the remark this answers is not in the document]
Comment (nrc) [reply src/part-guide/adv-async-await.md:59]: I would expect a comma here as that's the usual usage for `e.g.,` and we're not followed by a list, but by a figure

File: src/part-guide/adv-async-await.md (hunk 1 above)
Comment (nrc) [reply src/part-guide/adv-async-await.md:114]: Tracked for later...

File: src/part-guide/adv-async-await.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md]: Makes sense.

File: src/part-guide/adv-async-await.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md]: > Makes sense in general, in this case though I think the first 'in Rust' is warranted because I want this is a behaviour which is pretty unique to Rust and in contrast to other systems (and this is a known footgun for those coming from other languages).

Yes... when making the suggestion I'd almost gone the other way for that reason and suggested spelling that out instead.  Maybe that is the thing to do, is to add a sentence or two saying that outright.

> The second use is a bit of an in-joke referencing POJOs: https://en.wikipedia.org/wiki/Plain_old_Java_object

Ah, yes.  Well, carry on then.

File: src/part-guide/adv-async-await.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:59]: We use a colon for this elsewhere, e.g. in the Reference.  We'd say, "For example:" rather than "For example," when followed by a code block.  Or, "For example, consider this case:".

With the comma there, I'd expect the sentence to continue after the figure, or I'd expect the figure to end with a period (as is sometimes seen in mathematical writing).

The MLA seems [to follow](https://style.mla.org/colons-how-to-use-them/) this style, e.g.:

> Use colons to introduce a quotation when it is not integrated into the syntax of your sentence or otherwise requires a formal introduction:
>
> ...
>
> Use a colon to introduce the formal expression of a rule or principle:

Maybe more importantly, this is what Knuth [suggests](https://jmlr.csail.mit.edu/reviewing-papers/knuth_mathematical_writing.pdf) also (page 3):

> The statement just preceding a theorem, algorithm, etc., should be a complete sentence or should end with a colon.

File: src/part-guide/adv-async-await.md [hunk 5]
```diff
@@ -2,26 +2,143 @@
...
+- Via a [`CancellationToken`](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html) (which requires the future being canceled to notice the token and cooperatively cancel itself).
+- Implicitly, by a function or macro like [`select`](https://docs.rs/tokio/latest/tokio/macro.select.html).
+- By simply dropping a future if you own it.
+
+The first two are specific to Tokio, though most runtimes provide similar facilities. The second requires cooperation of the future being canceled, but the others do not. In these other cases, the canceled future will get no notification of cancellation and no opportunity to clean up (besides its destructor). Note that even if a future has a cancellation token, it can still be canceled via the other methods which won't trigger the cancellation token.
+
+From the perspective of writing async code (in async functions, blocks, futures, etc.), the code might stop executing at any `await` (including hidden ones in macros) and never start again. In order for your code to be correct (specifically to be *cancellation safe*), it must work correctly whether it completes normally or whether it terminates at any await point.
```
[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: Disagree.  Cancellation safety is analogous to and is as valid as unwind safety.  But first, let's check we mean the same thing by it.  When you say:

> we shouldn't cover "cancellation safety" as a good thing

I think to myself, "well of course it's a good thing."  All it means is that nothing bad happens if you cancel the future.  Certainly bad things happening when you cancel the future can't be a good thing.

This seemingly simple definition draws from [Lamport's 1977 paper](http://www.cis.umassd.edu/~hxu/courses/cis481/references/Lamport-1977.pdf), which is the body of  thought from which all of Rust's safety properties derive.  He says:

> A safety property is one which states that something will *not* happen.  For example, the partial correctness of a single process program is a safety property.  It states that if the program is started with the correct input, then it cannot stop if it does not produce the correct output.

The "bad things" characterization of this may be have introduced by Schneider, e.g. in 1982 ("Lecture Notes in Computer Science", Chapter 5: "Formal Foundation for Specification and Verification", Section 2: "Proving Safety Properties", page 221):

> A safety property is one that holds for an infinite sequence of states if and only if it holds for every finite prefix of it.  Such a property holds for a program if it holds for every sequence of states corresponding to an execution of that program; it will therefore be true throughout execution of the program.
>
> Intuitively, a safety property can be thought of as stating that something "bad" doesn't happen during execution.

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: > Cancellation safety is analogous to and is as valid as unwind safety.

While I agree that both properties are very similar, I'd be quick to note that unwind safety is broadly considered to be confusing, sees little usage outside libstd, and there have been repeated talk about deprecating it for those exact reasons.

Also I should probably point out that unwind safety has actually gone through the RFC process and is representable in the type system. While "cancellation safety" has not seen any RFCs and as-defined by Tokio is also unrepresentable in the type system.

> I think to myself, "well of course it's a good thing." All it means is that nothing bad happens if you cancel the future.

What you're (perhaps unintentionally) arguing for here is that it's actually good for APIs to violate the standard mechanism for cancellation that was designed as part of the `Future` API.

The question to ask here is: **what do we gain from breaking with the core Future design**? From my perspective (and this has been echoed by Tokio contributors I've discussed this with) the only real reason anyone practically cares about "cancellation safety" is if they want to use `select!`.

Back in 2018 the `select!` API perhaps seemed like a good idea. But six years and a lot of usage experience later, it seems pretty clear `select!` creates more problems than it solves.

I think we should take this opportunity to take the past six years of learnings and guide a new cohort of Rust developers to instead use structured APIs for concurrency and automatic propagation of cancellation.

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: We don't have to use Tokio's definition for this.  Their definition is a bit imprecise, in my view, which is why I draw the analogy with unwind safety.

If you leave a data structure in a state that violates one of your program invariants in the case of a panic, then you've violated unwind safety.  Similarly, if you leave a data structure in a state that violates one of your program invariants in the case that a future is polled and then dropped prior to resolving, then you've violated cancellation safety.

There are other ways to violate both unwind safety and cancellation safety, of course.

While we could of course deprecate the `AssertUnwindSafe` trait, that wouldn't make the concept of unwind safety go away.

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: > Cancellation is one of the most cited problems people have using async Rust. I believe we should take this opportunity to take the past six years of learnings and guide a new cohort of Rust developers to instead use structured APIs for concurrency and automatic propagation of cancellation.

Concretely, as it pertains to the book here, what is your suggested course of action in this regard?  What you write seems to suggest a broader proposed plan for the book than just not mentioning cancellation safety.

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: I think you're likely right we can define a form of "cancellation safety" that could be encoded in the type system. But I also feel like this conversation illustrates pretty well what I meant when I said this topic would probably do well as an appendix.

The definition of "cancellation safety" is not standardized, may never be standardized, and if it is it may still conflict with how for example Tokio uses the term. Like, this is guaranteed to be confusing for users who are just learning about this.

The only practical reason to cover "cancellation safety" early on is if we decide to center `select!` later on. So I'm proposing we don't do that, and instead discuss both separately outside the happy path.

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: > Concretely, as it pertains to the book here, what is your suggested course of action in this regard?

I believe we would benefit from centering systems,  consistent patterns, and compiler-checked APIs. This is much in line with [@tmandry's writing on making async Rust reliable](https://tmandry.gitlab.io/blog/posts/making-async-reliable/). For example:

> "That means deprecating `select` in favor of other combinators, like `merge`, […]"

In [a follow-up post](https://tmandry.gitlab.io/blog/posts/the-main-thing/) they also wrote the following:

> ## Correctness by construction
> 
> Great code is trivial to verify. It maintains its invariants by making invalid states unrepresentable. It leans on static checks to keep bugs out of the codebase and ease the burden on reviewers.

I strongly agree with the thrust of his writing: if we want (async) Rust to be usable by everyone, we have to work to make doing the wrong things impossible and the right things easy.

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: > The definition of "cancellation safety" is not standardized, may never be standardized, and if it is it may still conflict with how for example Tokio uses the term. Like, this is guaranteed to be confusing for users who are just learning about this.

This is a book.  It can give its own definition, as books often do, and as @nrc does here:

> From the perspective of writing async code (in async functions, blocks, futures, etc.), the code might stop executing at any `await` (including hidden ones in macros) and never start again. In order for your code to be correct (specifically to be *cancellation safe*), it must work correctly whether it completes normally or whether it terminates at any await point.

That's a perfectly cogent definition.  We don't necessarily need to care that Tokio's docs give a somewhat more narrow one.

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: > I believe we would benefit from centering systems, consistent patterns, and compiler-checked APIs.

Perhaps it'd be better for us to build this world first before trying to write a book about it.  That is, it seems like we wouldn't be well serving today's readers by describing something that's not how people generally write async Rust today.  Maybe this would be a better direction for the next iteration of this book.

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: > For me, this question doesn't turn on select!. E.g., any reasonable system is going to have a timeout(..) combinator.

How is that a practical example though? If a timeout occurs then that means you're no longer interested in the output of the future. If you're ready to ignore the timeout and resume the future: why would you put a timeout on it in the first place?

`select!` is unique in that it needs to treat cancellation as not being final. While retrying on timeout certainly is possible: that isn't done by ignoring cancellation and continuing with the same future. The only practical way anyone would do that is by calling the function again and creating a new future. Otherwise that's not retrying, that's just continuing.

---

> Perhaps it'd be better for us to build this world first before trying to write a book about it.

I mean: if all we're interested in is deferring to whatever established practices are, the Rust project would be best served by referring people directly to the Tokio book (or equivalent) and halt its own documentation efforts.

But I think there is value we can provide beyond that. Just like you feel that perhaps we can do better than Tokio's usage of the term "cancellation safety" — I too believe that we can improve on the status quo. So much so in fact that that's what I've been working on within the context of WG Async for the past few years. You seem to have missed this, but here are some of the libraries I've helped author over the years:

- [futures-concurrency](https://docs.rs/futures-concurrency) - a library created to replace all uses of `select!` - it provides a wide host of structured async concurrency APIs, none of which depend on "cancellation safety"
- [parallel-future](https://docs.rs/parallel-future) - a library that enables cancellation to propagate across task boundaries, authored to resolve a common problem we were seeing inside Microsoft. Absent async destructors there are some caveats however, which [I've documented here](https://blog.yoshuawuyts.com/tasks-are-the-wrong-abstraction/).
- [futures-time](https://docs.rs/futures-time/latest/futures_time/) - provides various time-based operations. This one is more experimental, but it exercises a number of the other future controls: not just cancellation, but also pausing and resuming. As well as provides a generalization of the "cancellation token" pattern.

Using futures in a way that is easy to use correctly, and hard to misuse is not a matter of building missing APIs - those already exist. It's mainly a matter of eduction, which is exactly what this book is about. We actually have a chance to educate users here on how to avoid most issues with futures today! Which is why I'm so strongly advocating for not placing the focus on `select!` and "cancellation safety" and instead centering structured concurrency.

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: Actually, let me zoom out a little here. We're likely talking past each other. @traviscross can you clarify your positions on the following:

- do you believe the book should cover `select!` on the happy path?
- do you believe the book should cover "cancellation safety" on the happy path?
- if you believe we should cover "cancellation safety" but not `select!` on the happy path, can you elaborate why you believe that is important for users to learn about?

*(By "happy path" I mean "not in an appendix or similar, but in the main text".)*

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: > You seem to have missed this, but here are some of the libraries I've helped author over the years... It's now mainly a matter of education...

Thanks.  I didn't miss these, but I also didn't realize that you considered these all to be in such a state so as to be ready for full scale production use and to have solved the ambitious set of problems that you set out.

[the remark this answers is not in the document]
Comment (nrc) [reply src/part-guide/adv-async-await.md:48]: I'll come back to the whole conversation later, but for now I want to point out (and/or confirm my understanding): cancellation safety is a think even if you only use drop to cancel and even if you never use `select`. That is, any future can be dropped and therefore cease execution at any await point, and you must either ensure that the future does not lose data when that happens, or ensure that it never happens. Not using `select` is necessary but not sufficient to ensure the latter.

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: > I didn't miss these, but I also didn't realize that you considered these all to be in such a state so as to be ready for full scale production use and to have solved the challenging set of problems that you set out above.

I mean, well, yeah — these libraries have been iterated on for the better part of a decade. They're basically all off-shoots of the `async-std` project. For example `futures-concurrency` now has over a million downloads now too, and is used in several high-visibility projects.

That's not to say all these libraries are perfect or anything. There is always more work to be done, and at the end of the day we're always limited by the language and standard library types. But yeah: I think it's fair to say most of these hit prime time a while ago.

I wouldn't go so far as to claim they solve all issues tho - we need language changes like async destructors for that. But I do believe that they allow us to solve most of the common issues people experience with async concurrency and cancellation. At a minimum they're a meaningful step up from the APIs we started off with six years ago.

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: > Cancellation safety is a think even if you only use drop to cancel and even if you never use select. That is, any future can be dropped and therefore cease execution at any await point, and you must either ensure that the future does not lose data when that happens, or ensure that it never happens.

I think you're getting at the disagreement between TC and me here. I assume TC would agree with you here that, yes, that's indeed what cancellation-safety is about: not losing any data in the face of cancellation.

I have a slightly different perspective here: I do not believe that a function or block halting execution on `.await` is meaningfully different from a function halting on `?`, `return`, `panic!`, ` break`, and so on. I believe it's worth covering these as a class of "things that may cause a function or scope to permanently cease making progress" rather than singling out `.await` as somehow being special.

Also Tokio uses a stricter definition of "cancellation safety" than you might expect. Their definition is closer to "freedom from side effects". As an example they name in their `select!` docs "changes in queue orderings" on a mutex as something that is not "cancellation safe".

I also believe that ignoring cancellation is  never a valid strategy for responding to cancellation. That is what leads to temporal logic errors such as dangling tasks that can cause data corruption or data loss, or the inability to retry operations when we actually intended to. Futures should handle the possibility of exiting on `.await` much the same as exiting on `?` or `return`.

The way to respond to cancellation should always be using destructors. In the future hopefully also using async destructors, which in turn would conveniently also set us up to encode true ["uncancellable futures"](https://sabrinajewson.org/blog/async-drop#uncancellable-futures) which would be free of the aforementioned temporal logic errors.

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: Trivially, this is a function that is not cancellation safe:

```rust
async fn self_destruct_selftest() {
    start_self_destruct_sequence(Duration::from_secs(10));
    sleep(Duration::from_secs(2)).await;
    abort_self_destruct_sequence();
}
```

We don't need `select!` to make this apparent, of course.  It's enough to say:

```rust
self_destruct_selftest().timeout(1).await;
//                      ~~~~~~~~~~~
//           Does not do what you might hope.
```

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: > I wouldn't go so far as to claim they solve all issues tho - we need language changes like async destructors for that.

Yes, and this is what I meant by building that future first.  There's much language work to be done, and I'm not sure to what degree we can reasonably point people toward patterns like structured concurrency ahead of doing that work.

> The APIs to use async Rust in a way that is easy to use correctly and hard to misuse already exist. It's now mainly a matter of eduction, which is exactly what this book is about. We actually have a chance here to educate users here on how to avoid most issues with futures today!

This is the other thing I must have missed.  It seems that you're essentially asking the project to endorse, at least in this book, these libraries as the currently-best solution to these problems.  I'm not saying that they're not, to be clear.  It just seems that that should be raised somewhere prominently.  @nrc's [stated plan](https://github.com/rust-lang/async-book/issues/224) for this book, e.g., was to:

> be practical, mostly use async/await, use Tokio, and avoid implementation details

This probably isn't how I'd write an async book for Rust the language.  I'd probably start at the bottom.  I'd avoid both Tokio and the libraries you mention, and I'd walk the reader though building an executor with only what's in the standard library and then work up from there.  As you say, the reader can read Tokio's guides for how to use Tokio.

But I'm not the one rewriting an async book here, so I'm willing to give a lot of latitude.

Anyway, at this point, I'm curious what the rest of @rust-lang/wg-async thinks.

[the remark this answers is not in the document]
Comment (traviscross) [reply src/part-guide/adv-async-await.md:48]: Also, cc @chriskrycho who had particular thoughts about how cancellation should be introduced [here](https://github.com/rust-lang/async-book/issues/224#issuecomment-2405983967).

[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/part-guide/adv-async-await.md:48]: > There's much language work to be done, and I'm not sure to what degree we can reasonably point people toward patterns such as structured concurrency ahead of doing that work.

The only limitation I'm aware of is that if we work with multi-threaded futures, while we can automatically propagate cancellation across thread boundaries. But we have no way to wait for that cancellation to finish without risking dead-locking on single-threaded runtimes. We need async destructors for this.

That's really all. I don't think that should affect our decision to tell people that e.g. future invocations should always have a parent future that `.await`s it. Or that people should use `Stream::merge` and probably avoid `select!`.

> It seems that you're essentially asking the project to endorse, at least in this book, these libraries as the currently-best solution to these problems and for how to build programs in async Rust.

I mean: a little bit, yeah. I don't think we need to cover `futures-time` here - but the APIs provided by `futures-concurrency` and `parallel-future` seem important enough. I assume that's also why `futures-concurrency` has been included in the outline of this chapter.

@chriskrycho's ask to thoroughly cover cancellation in this book is part of why I'm taking this position. Cancellation *is* incredibly important and something users struggle with. And I don't know how we can setup people for success without centering structured concurrency and explicitly steering people away from APIs that break the async cancellation model.

File: src/part-guide/adv-async-await.md [hunk 6]
```diff
@@ -2,26 +2,143 @@
...
+
+```rust,norun
+async {
+    let x = foo()?;   // This `?` only exits the async block, not the surrounding function.
+    consume(x);
+    Ok(())
+}.await?
+```
+
+Annoyingly, this often confuses the compiler since (unlike functions) the 'return' type of an async block is not explicitly stated. You'll probably need to add some type annotations on variables or use turbofished types to make this work, e.g., `Ok::<_, MyError>(())` instead of `Ok(())` in the above example.
```
Comment (eholk) [inline src/part-guide/adv-async-await.md:113]: I think it'd be good to make this example compile and run as is, rather than saying you'll probably need to modify it. Or does this one work and we got lucky, but in general you'll probably need a turbofish?

File: src/part-guide/adv-async-await.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (eholk) [reply src/part-guide/adv-async-await.md:48]: > Anyway, at this point, I'm curious to hear what the rest of @rust-lang/wg-async thinks.

Here are some of my thoughts.

This book should say _something_ about cancellation and cancellation safety. These terms are out there and if we didn't address them, it'd be surprising to readers. For example, someone might say "I see people talking all time on Reddit about cancellation safety. Why doesn't this book address it?"

The book has a lot of freedom to define these terms. It makes sense to be clear about what is meant and how that differs from Tokio's concepts of the same name (if they differ at all).

Similarly, for better or worse, `select!` is a big part of how most people experience async Rust. We should probably talk about it. I could see two approaches, mostly depending on how iconoclastic we want to be:

1. Write the book mainly using `select!`, with forward references saying something like "Structuring your code robustly using `select!` can be hard to get right. Check out Chapter NN on Structured Concurrency for another approach." Then, have a chapter that shows how structured concurrency really shines and why users should adopt that.
2. Write the book mainly using structured concurrency, noting that most of the ecosystem at the time of writing uses `select!` but we encourage the use of structured concurrency primitives because they are easier to get correct. Then add a chapter for `select!`, since if you're working with already-existing code, chances are you're going to need to understand `select!`.

I'd probably lean towards option 1 since I think that better reflects the state of the world today. Maybe for the second edition it'd be possible to structure the book around option 2.

I'm starting to think that cancellation safety might better be thought of as a design pattern or an API philosophy or discipline, rather than as a foundational safety property the way we think of memory safety or type safety. At least as I've been thinking about cancellation lately, it tends to be about programs such as:

```rust
let mut thing = Thagomizer::new();
thing.thagomize().timeout(ONE_SECOND).await;
thing.thagomize().await; 
```

Basically, we're concerned about the behavior if the timeout expires. What happens on the second call to `thagomize`? I don't know that we can make a blanket answer for all programs. The type system certainly cannot capture all of the behaviors and semantics of whatever you might put in place of `Thagomizer`. So at least for the foreseeable future, I think there's always going to be a place for documenting those properties we can't capture in code.

[the remark this answers is not in the document]
Comment (tmandry) [reply src/part-guide/adv-async-await.md:48]: I think the book should discuss the concept of writing code that is correct with respect to cancellation (execution ending at an await point), because it can be tricky to do right, and there have been a lot of bugs caused by code not handling this correctly.

I do think "cancellation safety" is a natural enough term for that, though there might be other terms that are better. It's useful to note, as others have, that other definitions exist than the one used in this book. It's valid to choose a different term for that reason, and maybe others have suggestions, but I also think when it comes to terminology we have an opportunity to "set the agenda" in this book.

The [particular definition](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety) @yoshuawuyts references on `tokio::select` is referring more to a property I might venture to call "restart safety", i.e. that an expression can be re-evaluated again after its original future was dropped, without undesirable side effects.[^misuse] This is a superset of the definition of cancellation safety used in this book, and as such I think we could reclaim the term and ask them to pick a more precise one.

[^misuse]: We don't have a good way of explaining this in terms of existing language concepts because `select` is a macro that extends the language with a new control-flow construct whose properties are unique and not widely understood. That is probably closely related to why it seems to be prone to misuse.

With respect to how we cover current tools versus emerging ones: I think we have to stick to the facts. In particular, we should list out the various patterns that one might encounter and the various ways of solving those patterns. We should explain that some solutions have known footguns, while others are newer and therefore less widely used and might be less general than the footgunny ones.

If we don't and try to protect users by not giving them the full picture, I think the async book never becomes relevant enough to have any influence. People need a resource that prepares them for the code that exists in the world that they need to understand, one that doesn't hide from them actual solutions they might need. It's okay to stick these in an "advanced" section with all the usual caveats, but while they are widely used we should still include them.

File: src/part-guide/adv-async-await.md (hunk 6 above)
Comment (nrc) [reply src/part-guide/adv-async-await.md:113]: This is just a snippet, whether it compiles or not depends on the surrounding function (and also the type of `foo`, probably).

File: src/part-guide/adv-async-await.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (nrc) [reply src/part-guide/adv-async-await.md:48]: I'm tracking this issue at https://github.com/rust-lang/async-book/issues/227#issuecomment-2496234560. I'll try and tighten up the language here a little for now, but intend to come back to this later. In particular, I'll have a think about how I want to talk about cancelation throughout the book.
